flocker/common/_retry.py

```diff
@@ -175,3 +178,39 @@ def poll_until(predicate, steps, sleep=None):
     if result:
         return result
     raise LoopExceeded(predicate, result)
+
+
+def retry_effect_with_timeout(effect, timeout, retry_wait=1, exp_backoff=True,
+                              time=time.time):
+    """
+    If ``effect`` fails, retry it until ``timeout`` expires.
+
+    To avoid excessive retrying, this function by default (it can be disabled)
+    uses the exponential backoff algorithm, waiting twice the time between
+    each retry.
+
+    :param Effect effect: The Effect to retry.
+    :param int timeout: Keep retrying until timeout.
+    :param int retry_wait: The wait time between retries
+    :param bool exp_timeout: Whether we should use exponential backoff
+    :param callable time: A nullary callable that returns a Unix UTC timestamp.
+    :return: An Effect that does what ``effect`` does, but retrying.
+    """
+
+    end_time = time() + timeout
+
+    def should_retry(e):
+        if time() > end_time:
+            return Effect(Constant(False))
+        else:
+            effect = Effect(Delay(should_retry.wait_secs)).on(
+                success=lambda x: Effect(Constant(True)))
+
+            if exp_backoff:
+                should_retry.wait_secs *= 2
+
+            return effect
+
+    should_retry.wait_secs = retry_wait
```

Comment:
Storing this on `should_retry` (i.e. assigning an attribute to a function) is a bit weird, but OK. No change required.